### PR TITLE
Check column access for view identities

### PR DESCRIPTION
**Pull request by rschlussel2**
Check column access for view identities.  The check was being skipped
when we moved column access checks to the Analyzer. With this change,
createViewWithSelectFromColumns can replace
createViewWithSelectFromTable/View

### CODE REVIEW

**Comment by rschlussel2** [comment]
ping @electrum

File: presto-main/src/main/java/com/facebook/presto/testing/TestingAccessControlManager.java [hunk 1]
```diff
@@ -313,7 +314,7 @@ public void checkCanCreateViewWithSelectFromView(TransactionId transactionId, Id
     public void checkCanCreateViewWithSelectFromColumns(TransactionId transactionId, Identity identity, QualifiedObjectName tableName, Set<String> columnNames)
     {
         if (shouldDenyPrivilege(identity.getUser(), tableName.getObjectName(), CREATE_VIEW_WITH_SELECT_COLUMNS)) {
-            denySelectView(tableName.toString());
+            denyCreateViewWithSelect(tableName.toString());
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/testing/TestingAccessControlManager.java]: Why change this but not the other `checkCanCreateViewXXX` methods?

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java [hunk 2]
```diff
@@ -925,21 +986,21 @@ public void testViewAccessControl()
...
                 "CREATE VIEW test_nested_view_access AS SELECT * FROM test_view_access",
-                privilege("test_view_access", CREATE_VIEW_WITH_SELECT_VIEW));
+                privilege("test_view_access", CREATE_VIEW_WITH_SELECT_COLUMNS));
 
         // verify selecting from a view over a view requires the view owner of the outer view to have special view creation privileges for the inner view
         assertAccessDenied(
                 "SELECT * FROM test_nested_view_access",
-                "Cannot select from view .*.test_view_access.*",
-                privilege(nestedViewOwnerSession.getUser(), "test_view_access", CREATE_VIEW_WITH_SELECT_VIEW));
+                "Cannot create view that selects from .*.test_view_access.*",
```
Comment (electrum) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java]: This seems like the wrong message for a SELECT query. Though this may be the best we can do given how permissions currently work.

**Comment by electrum** [comment]
Nit for second commit message: "DELETE clause" is uppercased but "where clause" is not

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java [hunk 3]
```diff
@@ -73,12 +74,14 @@ public Analysis analyze(Statement statement, boolean isDescribe)
...
 
         // check column access permissions for each table
-        analysis.getTableColumnReferences(session.getIdentity())
-                .forEach((tableName, columns) ->
-                        accessControl.checkCanSelectFromColumns(session.getRequiredTransactionId(),
-                                session.getIdentity(),
-                                tableName,
-                                ImmutableSet.copyOf(columns)));
+        analysis.getTableColumnReferences().forEach((identity, tableColumnReferences) -> {
+            AccessControl identityAccessControl = identity.equals(session.getIdentity()) ? accessControl : new ViewAccessControl(accessControl);
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java]: I'm not sure if this is a problem or just needs a comment....  The identity context could be `A->B->A`, and in that case the inner most `A` would be checked with a regular access control instead of a view access control.

Comment (electrum) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java]: This logic seems to be intended to be the same as in `StatementAnalyzer.Visitor#analyzeView`, where we do normal access checks for the view if the session user owns it, and I believe the logic should be the same. However, I think it's actually different because `StatementAnalyzer` is called recursively, and thus the session user used when analyzing the innermost `A` would be the owner of `B`. Here, the session user is constant across references.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 4]
```diff
@@ -391,7 +391,7 @@ protected Scope visitDelete(Delete node, Optional<Scope> scope)
                     session);
 
             Scope tableScope = analyzer.analyze(table, scope);
-            node.getWhere().ifPresent(where -> analyzer.analyzeWhere(node, tableScope, where));
+            node.getWhere().ifPresent(where -> analyzeWhere(node, tableScope, where));
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Can we make this change a separate commit? It seems to be an independent fix: enforce `SELECT` access checks for the `WHERE` clause.

File: presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorAccessControl.java [hunk 5]
```diff
@@ -195,6 +195,7 @@ default void checkCanSelectFromColumns(ConnectorTransactionHandle transactionHan
      *
      * @throws com.facebook.presto.spi.security.AccessDeniedException if not allowed
      */
+    @Deprecated
```
Comment (electrum) [inline presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorAccessControl.java]: Add a note to the Javadoc indicating what to use instead
```
@deprecated Use {@link #checkCanSelectFromColumns} instead

File: presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorAccessControl.java [hunk 6]
```diff
@@ -255,6 +257,7 @@ default void checkCanSelectFromView(ConnectorTransactionHandle transactionHandle
...
```
Comment (electrum) [inline presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorAccessControl.java]:
```
@deprecated Use {@link #checkCanCreateViewWithSelectFromColumns} instead

File: presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorAccessControl.java [hunk 7]
```diff
@@ -195,6 +195,7 @@ default void checkCanSelectFromColumns(ConnectorTransactionHandle transactionHan
...
     default void checkCanSelectFromTable(ConnectorTransactionHandle transactionHandle, Identity identity, SchemaTableName tableName)
     {
         denySelectTable(tableName.toString());
```
Comment (electrum) [inline presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorAccessControl.java]: This should delegate to `checkCanSelectFromColumns`. Then subclasses can delete their implementations of the deprecated methods after implementing the column level checks.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileBasedSecurity.java [hunk 8]
```diff
@@ -52,7 +52,7 @@ public void testAdminCanRead()
         queryRunner.execute(admin, "SELECT * FROM orders");
     }
 
-    @Test(expectedExceptions = RuntimeException.class, expectedExceptionsMessageRegExp = ".*Access Denied: Cannot select from table tpch.orders.*")
+    @Test(expectedExceptions = RuntimeException.class, expectedExceptionsMessageRegExp = ".*Access Denied: Cannot select from columns \\[clerk, orderkey, orderstatus, custkey, totalprice, comment, orderdate, orderpriority, shippriority\\] in table or view tpch.orders.*")
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileBasedSecurity.java]: We should keep the original message. Listing out all the columns due to table-level access denied is a bad user experience (especially for tables with hundreds of columns)

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java [hunk 9]
```diff
@@ -73,12 +74,14 @@ public Analysis analyze(Statement statement, boolean isDescribe)
...
+            tableColumnReferences.forEach((tableName, columns) ->
+                    identityAccessControl.checkCanSelectFromColumns(session.getRequiredTransactionId(),
+                            identity,
+                            tableName,
+                            ImmutableSet.copyOf(columns)));
```
[the remark this answers is not in the document]
Comment (rschlussel2) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java]: Good point.  Not sure why I had that.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 10]
```diff
@@ -82,7 +84,7 @@
     private final Map<NodeRef<Expression>, FieldId> columnReferences = new LinkedHashMap<>();
 
     // a map of users to the columns per table that they access
-    private final Map<Identity, Map<QualifiedObjectName, Set<String>>> tableColumnReferences = new LinkedHashMap<>();
+    private final Map<AccessControlInfo, Map<QualifiedObjectName, Set<String>>> tableColumnReferences = new LinkedHashMap<>();
```
[the remark this answers is not in the document]
Comment (rschlussel2) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: ok.   I'll create an enum to represent access type.  Values are `ALLOW/VIEW/DEFAULT`.   

`select` and `delete` would be misleading names because not all select queries would use the `select` access control (views).  And lots of other privileges would also use it (including delete for some things).

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java (hunk 3 above)
Comment (rschlussel2) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java]: I'm combining the delete and view commits, so basically we explicitly state what access type we're dealing with and use the appropriate access control.  That should fix this issue.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (electrum) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: I'm not sure this makes sense from an analysis standpoint. We want to report facts about the type of column reference, not how they should be access checked later. The type of access can be translated into an access check policy, but shouldn't be specific to access control.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 11]
```diff
@@ -614,22 +616,24 @@ public JoinUsingAnalysis getJoinUsing(Join node)
         return joinUsing.get(NodeRef.of(node));
     }
 
-    public void addTableColumnReferences(Identity identity, Multimap<QualifiedObjectName, String> tableColumnMap)
+    public void addTableColumnReferences(AccessControl accessControl, Identity identity, Multimap<QualifiedObjectName, String> tableColumnMap)
     {
-        Map<QualifiedObjectName, Set<String>> references = tableColumnReferences.putIfAbsent(identity, new LinkedHashMap<>());
+        AccessControlInfo accessControlInfo = new AccessControlInfo(accessControl, identity);
+        Map<QualifiedObjectName, Set<String>> references = tableColumnReferences.computeIfAbsent(accessControlInfo, (k) -> new LinkedHashMap<>());
```
[the remark this answers is not in the document]
Comment (rschlussel2) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: can't use LinkedHashMap::new because it can't resolve the constructor (the function takes an argument that we don't used, and there are single argument constructors for LinkedHashMap)

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 12]
```diff
@@ -614,22 +616,24 @@ public JoinUsingAnalysis getJoinUsing(Join node)
...
                 .forEach((key, value) -> references.computeIfAbsent(key, k -> new HashSet<>()).addAll(value));
     }
 
-    public void addEmptyColumnReferencesForTable(Identity identity, QualifiedObjectName table)
+    public void addEmptyColumnReferencesForTable(AccessControl accessControl, Identity identity, QualifiedObjectName table)
     {
-        tableColumnReferences.putIfAbsent(identity, new LinkedHashMap<>());
-        tableColumnReferences.get(identity).putIfAbsent(table, new HashSet<>());
+        AccessControlInfo accessControlInfo = new AccessControlInfo(accessControl, identity);
+        tableColumnReferences.putIfAbsent(accessControlInfo, new LinkedHashMap<>());
```
[the remark this answers is not in the document]
Comment (rschlussel2) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: done, except not using LinkedHashMap/HashSet::new for reason explained above.

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java [hunk 13]
```diff
@@ -892,7 +896,7 @@ public void testViewAccessControl()
         // verify selecting from a view over a table requires the view owner to have special view creation privileges for the table
         assertAccessDenied(
                 "SELECT * FROM test_view_access",
-                "Cannot select from table .*.orders.*",
+                "Cannot create view that selects from .*.orders.*",
```
[the remark this answers is not in the document]
Comment (electrum) [reply presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java]: Looks good, though make sure to start the error message with a capital letter
